Mathlib/Dynamics/TopologicalEntropy/Semiconj.lean

```diff
@@ -0,0 +1,197 @@
+  rw [mem_ball_symmetry (V_symm.dynEntourage T n)] at x_i gs_cover
+  exact ⟨x_i, gs_cover⟩
+
+lemma le_coverMincard_image (h : Semiconj φ S T) (F : Set X) {V : Set (Y × Y)}
+    (V_symm : SymmetricRel V) (n : ℕ) :
+    coverMincard S F ((Prod.map φ φ) ⁻¹' (V ○ V)) n ≤ coverMincard T (φ '' F) V n := by
+  rcases eq_top_or_lt_top (coverMincard T (φ '' F) V n) with h' | h'
+  · exact h' ▸ le_top
+  rcases (coverMincard_finite_iff T (φ '' F) V n).1 h' with ⟨t, t_cover, t_card⟩
+  rcases t_cover.preimage h V_symm with ⟨s, s_cover, s_card⟩
+  rw [← t_card]
+  exact s_cover.coverMincard_le_card.trans (WithTop.coe_le_coe.2 s_card)
+
+lemma coverMincard_image_le (h : Semiconj φ S T) (F : Set X) (V : Set (Y × Y)) (n : ℕ) :
+    coverMincard T (φ '' F) V n ≤ coverMincard S F ((Prod.map φ φ) ⁻¹' V) n := by
+  rcases eq_top_or_lt_top (coverMincard S F ((Prod.map φ φ) ⁻¹' V) n) with h' | h'
+  · exact h' ▸ le_top
+  · rcases (coverMincard_finite_iff S F ((Prod.map φ φ) ⁻¹' V) n).1 h' with ⟨s, s_cover, s_card⟩
+    rw [← s_card]
+    rcases s_cover.image h with ⟨t, t_cover, t_card⟩
+    exact t_cover.coverMincard_le_card.trans (WithTop.coe_le_coe.2 t_card)
+
+open ENNReal EReal Filter
+
+lemma le_coverEntropyEntourage_image (h : Semiconj φ S T) (F : Set X) {V : Set (Y × Y)}
+    (V_symm : SymmetricRel V) :
+    coverEntropyEntourage S F ((Prod.map φ φ) ⁻¹' (V ○ V)) ≤ coverEntropyEntourage T (φ '' F) V :=
+  (limsup_le_limsup) (Eventually.of_forall fun n ↦ (monotone_div_right_of_nonneg
+    (Nat.cast_nonneg' n) (log_monotone (ENat.toENNReal_mono (le_coverMincard_image h F V_symm n)))))
+
+lemma le_coverEntropyInfEntourage_image (h : Semiconj φ S T) (F : Set X) {V : Set (Y × Y)}
+    (V_symm : SymmetricRel V) :
+    coverEntropyInfEntourage S F ((Prod.map φ φ) ⁻¹' (V ○ V))
+    ≤ coverEntropyInfEntourage T (φ '' F) V :=
+  (liminf_le_liminf) (Eventually.of_forall fun n ↦ (monotone_div_right_of_nonneg
+    (Nat.cast_nonneg' n) (log_monotone (ENat.toENNReal_mono (le_coverMincard_image h F V_symm n)))))
+
+lemma coverEntropyEntourage_image_le (h : Semiconj φ S T) (F : Set X) (V : Set (Y × Y)) :
+    coverEntropyEntourage T (φ '' F) V ≤ coverEntropyEntourage S F ((Prod.map φ φ) ⁻¹' V) :=
+  (limsup_le_limsup) (Eventually.of_forall fun n ↦ (monotone_div_right_of_nonneg
+    (Nat.cast_nonneg' n) (log_monotone (ENat.toENNReal_mono (coverMincard_image_le h F V n)))))
+
+lemma coverEntropyInfEntourage_image_le (h : Semiconj φ S T) (F : Set X) (V : Set (Y × Y)) :
+    coverEntropyInfEntourage T (φ '' F) V ≤ coverEntropyInfEntourage S F ((Prod.map φ φ) ⁻¹' V) :=
+  (liminf_le_liminf) (Eventually.of_forall fun n ↦ (monotone_div_right_of_nonneg
+    (Nat.cast_nonneg' n) (log_monotone (ENat.toENNReal_mono (coverMincard_image_le h F V n)))))
+
+/-- The entropy of `φ '' F` equals the entropy of `F` if `X` is endowed with the pullback by `φ`
+  of the uniform structure of `Y`.-/
+theorem coverEntropy_image (u : UniformSpace Y) {S : X → X} {T : Y → Y} {φ : X → Y}
```

Comment:
In most situations, I expect that people will already have a uniform structure on the space above, so maybe it's worth emphasizing in the lemma name that it's not designed for this case. Maybe rename it `coverEntropy_image_of_comap`?
Same thing for the next lemma.

Reply:
Done.